src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -1369,11 +1561,14 @@ bool rasterClear(SwSurface* surface)
 bool rasterImage(SwSurface* surface, SwImage* image, const Matrix* transform, const SwBBox& bbox, uint32_t opacity)
 {
     Matrix invTransform;
-    auto isUpScaling = false;
+    float scaling = 1.0f;
 
     if (transform) {
         if (!_inverse(transform, &invTransform)) return false;
-        isUpScaling = (transform->e11 * transform->e11) + (transform->e21 * transform->e21) > 1 ? true : false;
+        scaling = sqrt((transform->e11 * transform->e11) + (transform->e21 * transform->e21));
```

Comment:
since we know the scaling factor here, I'd calculate here `halfInvScaling = 0.5f/scaling` and pass this value to the down scaling functions as the `halfScaleRatio` value. In this way we don't need to call sqrt and multiplications. seams to be cheaper. what do u think?